Feature/slo check

linkous8 opened the pull request.

linkous8 committed:
WIP slo check mode implemented

linkous8 committed:
SLO check mode tested
- Allow nr api env var config
- parse slo metric vals into decimal instead of default float
- remove test print
- forbid malformed condition
- switch to yaml safe loader
- move progress msg after slo check
- update slo check results to be json serializable

linkous8 committed:
Implemented fast_fail measure
- Handle newrelic returning no metrics for instance
- Add counters for slo check failure
- Added passthrough sleep func break up sleeps, interweave slo checks
- Return rejected status when condition `for` met

linkous8 committed:
Change slo status aborted, include times in msg

leokalev commented on measure:
```diff
@@ -87,6 +100,113 @@ insights_session = url_session(INSIGHTS_API_URL.format(account_id=NEWRELIC_ACCOU
+    metrics: Dict[str, Any],
+    result: Dict[str, Any],
+    is_threshold: bool,
+) -> bool:
```
(cosmetic)
wouldn't it be simpler to define this as (metric: Dict) -> (result:float, err:string)?

(rather than pass metrics, metric_name and a boolean that defines in an indirect way where to store the output in a 'result' dictionary)

Providing the name might have been useful, if it were used in preparing error messages, but it isn't (except for the check that name is present in metrics).

If this is for the purpose of 'constructing' error messages from bits and pieces, this is a practice that should be avoided (mind, I was guilty of it myself in the past on many occasions). Even if this stays as an English-only code, future changes can easily result in funny status messages. And if localization is ever needed, this will be rejected by translators as unworkable. It is best to use entire sentences and have replacement strings in them only for values that come from user data that must be quoted-as is, or are themselves entire sentences that can always fit into another sentence as a subordinate clause, such as "an error has occurred in computing FOO: {placehoder-for-error-sentence-here}".

BTW, same goes for using the python-expression-inside-a-text-string thingie (recently imported into Python directly from Perl hell). It is not something that you want people that don't read Python (such as proof-readers or translators) to see or mess with. IMO, "{{compute_this('mess')}" should not be used to format text intended for non-Pythonized humans.

Please consider changing the function signature and moving the "error context" formatting to the caller, such as "error in aggregating metric {stuff}': {error-text-from-aggregate-here}".

(like I said, not critical or urgent, functionality-wise this is fine)

linkous8 replied:
The result also [stores the relevant instance IDs](https://github.com/opsani/servo-newrelic/blob/48e0b9c49000134c7231952b1af667297a99acdc/measure#L140) that returned data for the query. This is printed to the driver's STDOUT in the full results payload which is functionally ignored by servo (no 'progress' key and will never be the last stdout line). However, if servo is set to verbose slo check results will show up in the servo container logs for debugging.

If this is overkill, I could remove the instance_ids storing from the result. Or I could have the inst IDs come back as a third field of the tuple (if tuples are larger than 2 I generally lean toward return by reference). LMK your thoughts.

WRT: message clause constructing, I hadn't considered your points and really appreciate the feedback. I have a cleanup branch where I nitpicked all the bad style I could find and the message construction is more in line with your suggestion but there are two lines that may still be too computation heavy. If time permits, could you give those lines a quick look and LMK if they could benefit from being organized differently?

- https://github.com/opsani/servo-newrelic/blob/dae35761ea0db8e4ff0fd3fb7a13d56251438e66/slo.py#L216
- https://github.com/opsani/servo-newrelic/blob/dae35761ea0db8e4ff0fd3fb7a13d56251438e66/slo.py#L182-L186

leokalev commented on measure:
```diff
@@ -87,6 +100,113 @@ insights_session = url_session(INSIGHTS_API_URL.format(account_id=NEWRELIC_ACCOU
+
+    metric_values = metrics[metric_name]['values']
+    if not metric_values:
+        result['message'] = f'{preamble} metric contained no instances or values'
```
(cosmetic) the 'if' actually checks for 'metric contained empty list of instances'

leokalev commented on measure:
```diff
@@ -87,6 +100,113 @@ insights_session = url_session(INSIGHTS_API_URL.format(account_id=NEWRELIC_ACCOU
+            continue
+
+        # Pull metric values from timestamps, filter out falsey values; 0, 0.0
+        single_inst_vals = [decimal.Decimal(v[1]) for v in val_list if v[1]] # TODO: NaN check?
```
Why are zero values simply skipped over? AFAIK, metrics may legitimately have zero value (e.g., error count). 

Not even a negative value should be assumed invalid or impossible: there may well be measurable things that have a valid range like -1000 ... 1000, or similar.

(rant: "Falsey" is a word? I must be getting old. As is my browser's spell checker - it highlighted this as a typo. Anyway, metrics are by definition numeric, might as well use 'zero value' in the message text)

linkous8 replied:
Per the very last bullet point of the spec doc under Unresolved Issues, discerning 0 values from missing values is still being worked out. I did some digging and found if there is no data for a metric in a given timeslice, NewRelic returns 0.  Per the [docs](https://docs.newrelic.com/docs/apis/rest-api-v2/basic-functions/extract-metric-timeslice-data/#non-existent_data_help)

> Metric values that have existed in the past, but are no longer collected, will return a zero value.

I was also able to find some discussion on this in the community hub. This [particular issue](https://discuss.newrelic.com/t/api-returns-values-of-0-instead-of-null-when-using-average-function-on-a-time-period-with-no-data-points/6467/3?u=ben108) presents a solution in checking for a metric that should never be zero and using that to differentiate between missing data and true zero. In our case, we could define a `sentinel_metric` configuration within applicable Conditions for which true zero is a value that we must include in the calculations

https://discuss.newrelic.com/t/feature-idea-timeseries-queries-with-sparse-data-points/34591/13
https://discuss.newrelic.com/t/apm-layers-in-graphs-take-sharp-plunges-whilst-waiting-on-fresh-data/31835
https://discuss.newrelic.com/t/time-series-no-data-instead-of-0/45807

leokalev reviewed (commented):
Please check the note about zero value skipping, it seems suspect (see inline comment)

I did not see anything else that may need change before this is used.

Notes not related to the proposed 'slo check' change:
class docstring in class MeasureError(): "base class for error exceptions defined by drivers."
somewhat incorrect. This *is* the driver, not something defined 'for drivers'. (seems like this definition was meant to be in the base-class library)

There is no README.md (I looked it when I noticed that I could not immediately guess the meaning of the 'delay' config setting, it is not obvious at all what it does or why it's even needed). There is no module doc string, either - in other words, if you want to configure it, you have to Use the Source.

leokalev commented on measure:
```diff
@@ -247,48 +380,102 @@ class NewRelicDriver(Measure):
-        self.progress_message = 'Measurements started'
-        self.print_progress()
+        sleep_fn = time.sleep # overriden when slo provided
+        skip_sleep = os.environ.get('SKIP_SLEEP', False) # NOTE: this skips SLO checking as well as sleeps
```
I assume that 'SKIP_SLEEP' is used only in debugging/testing, the code doesn't seem to be operable without it for any normal measurement. I recommend adding a comment that this is a debug-only setting (or, renaming the setting to something like DEBUG_SKIP_SLEEP might help, in lieu of an explanatory comment).

linkous8 replied:
Renamed skip_sleep to DEBUG_SKIP_SLEEP, moved to the top for visibility and capitalized to reflect the fact its not changed. Added a comment to indicate it should be used only for debug and not in optimizations

leokalev commented on measure:
```diff
@@ -247,48 +380,102 @@ class NewRelicDriver(Measure):
+        slo_check = control.get('userdata', {}).get('slo_check')
+        slo = control.get('userdata', {}).get('slo')
+
+        if slo_check == 1 and slo is None:
```
(cosmetic?)

the 'slo_check' var is used in this exact construct (slo_check == 1) and only in this construct. Maybe, set it initially as "slo_check = get(userdata, ...) == 1" and then use it as a proper boolean.

NOTE that this construct means 'slo_check=1' enables SLO check, any other value disables it (e.g, slo_check=2, or slo_check="yes" would both disable it). Is this the actual intent?

(same question for fast_fail_config.disabled != 1 elsewhere in the code)

linkous8 replied:
For slo_check, I added a [simple validator](https://github.com/opsani/servo-newrelic/blob/48e0b9c49000134c7231952b1af667297a99acdc/measure#L395-L396) since the list of possible values is quite small. 

`fast_fail_config.disabled` is validated by a [pydantic conint](https://github.com/opsani/servo-newrelic/blob/48e0b9c49000134c7231952b1af667297a99acdc/measure#L177) (I don't do the same for slo_check because it seemed like overkill to define a pydantic class to parse a single flag)

leokalev commented on measure:
```diff
@@ -416,6 +601,121 @@ class NewRelicDriver(Measure):
+        }
+
+
+    def slo_check_sleep(self, secs: float, slo_input: SloInput) -> None:
```
why not check_slo_sleep()? (to match the related check_slo() function) - or do them both the other way (slo_check() and slo_check_sleep()).

linkous8 replied:
slo_check_sleep -> check_slo_sleep

linkous8 committed:
PR feedback changes
- Clarify impact of skip_sleep behavior
- Clarify no instances message
- Metrics always numeric, fix comment and call out no data
- Add validation for slo_check flag value, simplify bool exprs
- Rename slo_check_sleep for naming consistency

linkous8 committed:
Remove := assign to maintain py<3.8 compatibility

linkous8 committed:
Add raise_for_status to http responses

benburdick reviewed (approved):
Tested and ready to merge

linkous8 commented:
Per @snarfbeest on slack

>other than the zero->nothing item, there should be no reason not to start using the code as-is. I understand this is a known issue and assume that the proposed resolution is to omit all zeros as 'no data' for now and implement the additional processing later - and you are already tracking this as a to-do item. 

>I see you've marked part of the cosmetic/maintainability items as resolved - you can fix whatever makes sense from the remaining ones, time permitting, but none of them change functionality so it is fine to cut a release without them.
